### ISSUE TITLE
(autokey-qt) CPU and RAM usage goes up when re-log in to Desktop

### ISSUE THREAD

**Issue by nemonein** [autokey/autokey#198]
## Classification:

Bug?

## Reproducibility:

Always

## Summary

CPU usage goes up to 25%
RAM usage goes up to 85%

## Steps to Reproduce (if applicable)

Kubuntu 18.04, Autokey-qt 0.95.4
Run autokey-qt, and log out of the KDE/ log in again.
When log in, "AutoKey is already running as pid 1553 but is not responding" message is pops up, after I click 'OK' button, CPU/Ram usage goes up to 25/85% each.

## Expected Results

- Normally, CPU % is 0, RAM usage is about 40MiB.

## Actual Results

- PC lags..

## Version

0.95.4

Used GUI (Gtk, Qt, or both): qt on KDE

If the problem is known to be present in more than one version, please list all of those.
At least from 0.95.3.

Installed via: Deb package from here.

Distro: Kubuntu 18.04

## Notes

Ram usage for Autokey-qt is goes up to 85%(13.2GiB/15.6GiB), PC lags, I could barely kills the process.
CPU usage is somewhat stays at 25%.

**Comment by luziferius** [comment]
Hmm. That message indicates that 2 instances are started in quick succession. Normally, the second instance should exit gracefully. But that might fail, if the first hasn't set up the DBus interface when the second one starts.

Either disable the autostart option and let the KDE session restoration magic handle the application start, or exclude AutoKey from the restoration mechanic.

I'll try to reproduce this later and see if i can get the second process to actually exit in this case without hogging all RAM. It seems that it enters some infinite loop after that message dialogue is confirmed...

**Comment by josephj11** [comment]
No idea if this is related, but forever (back to 0.90.4), there are times when AutoKey  starts using a ton of CPU. Sometimes, it might be due to something in one of my scripts getting into a loop. I don't know. Killing AutoKey and restarting it "fixes" it. I have no idea how to intentionally recreate it.

**Comment by luziferius** [comment]
@nemonein I tried to reproduce this on my side and failed.
The dialogue window has a details view containing the exact error message, please post it?

Whenever I provoke it, the process exits as expected after I close the error message window.
So two things
- I’d like to have the message from the dialog details.
- Does the process start hogging resources only after you close the dialogue?

If yes, please let the dialogue window be open the next time it happens. Open KSysGuard/the system monitor. In the process table, enable the `Command` column (right click on the header to enable it). Filter for `autokey-qt`. Are there two instances? One should have additional arguments (like `-session`). Are both idle?
Then close the dialogue. Does one instance exit?
If it starts to hog resources, you can use the system monitor to kill it.

Try patching your installation: (The file is located at `/usr/lib/python3/dist-packages/autokey/qtapp.py`)
https://github.com/autokey/autokey/blob/385352ccf0f1075cbc36be73d059f5730f7c0257/lib/autokey/qtapp.py#L149-L152
Comment out or remove line 151. Removing it will remove the error dialog window and just exit the process.
Does this help?

**Comment by nemonein** [comment]
Thanks.

I want to let you know first that I turned on `Start with an empty session` on KDE System settings.

Here are my answers for your question:
The error message is,
```
 org.freedesktop.DBus.Error.ServiceUnknown: The name org.autokey.Service was not provided by any .service files
```
and, the process starts hogging resources even before I close the window.

There are 2 processes. I hope this screen shot might help you. (Sorry for Korean letters, but you might guess what it means.)
![autokey-1](https://user-images.githubusercontent.com/5688288/47089684-52a26a00-d25c-11e8-98dc-c98a1815d912.png)

If I close the dialog, one instance(this is the new instance) exits. However the old instance still lives and hogs the resources.

I commented out line number 151 on qtapp.py, but nothing changes.

...........................................
I've tested this on different machines, and the results are different.

- Desktop PC(Kubuntu 18.04 with Backports PPA enabled, 16GB Ram) - The same as above.
- Laptop PC(Kubuntu 18.04 with Backports PPA enabled, 8GB Ram) - The same as above.
- VirtualBox (Kubuntu 18.04 w/o Backports PPA, 2GB Ram) - 2 Processes stays together, new one opens a dialog box(Error message), after Desktop has been restarted(re login), the old process exits within several minutes.
- Laptop PC(KDE Neon, Kubuntu 18.04 based) - Works without any problem. The Autokey process exits when I log out of the desktop. So, on re-login, there's only one Autokey instance.

I hope you could find something with these.

**Comment by luziferius** [comment]
Thanks you. Korean text is not a problem. I don’t have to read the text to get the content^^

- I assumed both instances are started at login and the exit logic is somehow broken. You confirmed it is not.
- It looks like it may be caused by the backports PPA. I’ll investigate using virtualbox, when I have some time to do it.
- So the autokey process does not exit when you log out or end your desktop session?

To verify the last point:
- Log in to your GUI session, verify that a working autokey instance is running.
- Log out again
- Log in on a terminal session (like on TTY2, available using [Ctrl]+[Alt]+[F2]), and run `systemctl status` and `ps -x | grep "autokey" | grep -v "grep"`. Does the autokey instance still run?

If so, your session log-out is broken somehow and does not kill all running GUI programs.

Another thing you can try:
- Exit autokey.
- Start autokey again using `autokey-qt --verbose > log_file.log` & This will write a full debug log to `log_file.log`. It contains all pressed keys and other debug log entries
- Then provoke the issue.
- Kill autokey, after it started to hog resources.
- Verify that the log does not contain sensitive data. You don’t want to publish your passwords, if you typed any.
- Then upload the log file. (you can drop files into the github text editor.)

I hope that the log reveals where autokey leaks memory. Then I can try fixing the memory leak.

**Comment by nemonein** [comment]
Thank you for your effort.
I'll try later, but there are something I can tell you right now.

There's no sign of it on the picture above, but the pid for old process is the same as the one that I logged in before.
Here's what I mean.

- For example, the pid of autokey-qt is 1658. Then log out.
- Log in again, there two autokey-qt instances, one(which is hogging) is 1658, and another one is 1700.

About the Backports PPA, it was my mistake. The VB's Kubuntu does not have Backports enabled. So, I don't think Backports did anything for this issue.

That thing you've asked, I'll test it and report later.

**Comment by luziferius** [comment]
> There's no sign of it on the picture above, but the pid for old process is the same as the one that I logged in before.

Ok. You can skip the middle part of my last post. You just confirmed that the process survives the logout.

**Comment by nemonein** [comment]
Here's the report.
```
$ ps -x | grep "autokey" | grep -v "grep"
 1351 ?        Sl     0:00 /usr/bin/python3 /usr/bin/autokey-qt
```
It is running, while I log in again on to a terminal session only.

And here the log file goes.
[log_file.zip](https://github.com/autokey/autokey/files/2491966/log_file.zip)
There's no sensitive data in it.

**Comment by luziferius** [comment]
Thanks, the log is useful.

It goes into an infinite loop when it tries to empty the event buffer.
But each time it tries, it fails (because the display connection is closed) and re-tries without ever exiting the loop.

For reference, this is the begin of the loop:
```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/autokey/interface.py", line 900, in __flushEvents
    for x in range(self.localDisplay.pending_events()):
  File "/usr/lib/python3/dist-packages/Xlib/display.py", line 192, in pending_events
    return self.display.pending_events()
  File "/usr/lib/python3/dist-packages/Xlib/protocol/display.py", line 244, in pending_events
    self.send_and_recv(recv = 1)
  File "/usr/lib/python3/dist-packages/Xlib/protocol/display.py", line 610, in send_and_recv
    raise self.socket_error
Xlib.error.ConnectionClosedError: Display connection closed by server
2018-10-18 21:24:34,152 ERROR - interface - __flushEvents: Some exception occured:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/autokey/interface.py", line 894, in __flushEvents
    readable, w, e = select.select([self.localDisplay], [], [], 1)
  File "/usr/lib/python3/dist-packages/Xlib/display.py", line 157, in fileno
    return self.display.fileno()
  File "/usr/lib/python3/dist-packages/Xlib/protocol/display.py", line 185, in fileno
    self.check_for_error()
  File "/usr/lib/python3/dist-packages/Xlib/protocol/display.py", line 352, in check_for_error
    raise err
  File "/usr/lib/python3/dist-packages/autokey/interface.py", line 900, in __flushEvents
    for x in range(self.localDisplay.pending_events()):
  File "/usr/lib/python3/dist-packages/Xlib/display.py", line 192, in pending_events
    return self.display.pending_events()
  File "/usr/lib/python3/dist-packages/Xlib/protocol/display.py", line 244, in pending_events
    self.send_and_recv(recv = 1)
  File "/usr/lib/python3/dist-packages/Xlib/protocol/display.py", line 610, in send_and_recv
    raise self.socket_error
Xlib.error.ConnectionClosedError: Display connection closed by server
2018-10-18 21:24:34,152 ERROR - interface - __flushEvents: Some exception occured:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/autokey/interface.py", line 894, in __flushEvents
    readable, w, e = select.select([self.localDisplay], [], [], 1)
  File "/usr/lib/python3/dist-packages/Xlib/display.py", line 157, in fileno
    return self.display.fileno()
  File "/usr/lib/python3/dist-packages/Xlib/protocol/display.py", line 185, in fileno
    self.check_for_error()
  File "/usr/lib/python3/dist-packages/Xlib/protocol/display.py", line 352, in check_for_error
    raise err
  File "/usr/lib/python3/dist-packages/autokey/interface.py", line 894, in __flushEvents
    readable, w, e = select.select([self.localDisplay], [], [], 1)
  File "/usr/lib/python3/dist-packages/Xlib/display.py", line 157, in fileno
    return self.display.fileno()
  File "/usr/lib/python3/dist-packages/Xlib/protocol/display.py", line 185, in fileno
    self.check_for_error()
  File "/usr/lib/python3/dist-packages/Xlib/protocol/display.py", line 352, in check_for_error
    raise err
  File "/usr/lib/python3/dist-packages/autokey/interface.py", line 900, in __flushEvents
    for x in range(self.localDisplay.pending_events()):
  File "/usr/lib/python3/dist-packages/Xlib/display.py", line 192, in pending_events
    return self.display.pending_events()
  File "/usr/lib/python3/dist-packages/Xlib/protocol/display.py", line 244, in pending_events
    self.send_and_recv(recv = 1)
  File "/usr/lib/python3/dist-packages/Xlib/protocol/display.py", line 610, in send_and_recv
    raise self.socket_error
```
Each time the error handler tries to handle the error, another exception is risen and the loop continues. Each time, the stack is appended and grows, as you see. This causes the memory consumption.

**Comment by nemonein** [comment]
It's my pleasure to help you.
Thank you again.

**Comment by luziferius** [comment]
I tried to fix the issue and attached a ZIP archive containing updated debian packages. (You initially said you used the debs I uploaded, so I think this should be the easiest thing.)

[potential_fix.zip](https://github.com/autokey/autokey/files/2492056/potential_fix.zip)

You’ll install them by using `dpkg --install autokey-common_0.95.4-1_all.deb autokey-qt_0.95.4-1_all.deb`

The potential fix included in the packages can be seen here: https://github.com/autokey/autokey/commit/96f03f4390c25e45e7f6bf9dea9105fd38007e4c
When a ConnectionClosedError is encountered, autokey now explicitly exits the event loop.

**Comment by nemonein** [comment]
I've installed the packages. (By the way, `sudo apt install x.deb` also works.)

The error occurs again, but no resource hogging.
```
$ ps -x | grep "autokey" | grep -v "grep"     
 1642 ?        Sl     0:01 /usr/bin/python3 /usr/bin/autokey-qt
 2485 ?        Sl     0:00 /usr/bin/python3 /usr/bin/autokey-qt
```
1642 is the prior one, and 2485 is the new one.
![autokey-3](https://user-images.githubusercontent.com/5688288/47158766-4f27e500-d327-11e8-918b-14350e3b7216.png)

I closed the error dialog box by clicking ok, and 2485 exits, but 1642 still remains.
However, autokey does not work. Not on the tray, no response at all.

**Comment by luziferius** [comment]
At least it does not crash the system anymore, so its somehow an improvement…

Can you please create a new log file, like you already did?

**Comment by nemonein** [comment]
Sure, wait for a minute.

**Comment by nemonein** [comment]
2nd log.
[log_file_20181057.zip](https://github.com/autokey/autokey/files/2492176/log_file_20181057.zip)

**Comment by luziferius** [comment]
I think I requested unneeded work from you… I looked at the log again. The exit logic is never called. I think I know how to fix this.
So instead of exiting the event loop, It should additionally call the shutdown() function.

**Comment by nemonein** [comment]
Never mind. I'm glad you get a clue.

**Comment by luziferius** [comment]
Let’s try if calling Application.shutdown() helps, when it encounters a ConnectionClosedError.
[autokey_0.95.4-2.zip](https://github.com/autokey/autokey/files/2492273/autokey_0.95.4-2.zip)
Commit: 
https://github.com/autokey/autokey/commit/36cbe8cb397371757526b25f499dd5ff4c791477

**Comment by nemonein** [comment]
I'm sorry.
Still 2 instances.
```
$ ps -x | grep "autokey" | grep -v "grep"
 1423 ?        Sl     0:00 /usr/bin/python3 /usr/bin/autokey-qt
 2139 ?        Sl     0:00 /usr/bin/python3 /usr/bin/autokey-qt
```

**Comment by copyme** [comment]
@luziferius I have applied your patch and so far things look OK. I will continue to test this. BTW, you can get RPMs with the patch from here: h[ttps://build.opensuse.org/package/binaries/home:copyme:branches:X11:common:Factory/autokey/openSUSE_Factory ](ttps://build.opensuse.org/package/binaries/home:copyme:branches:X11:common:Factory/autokey/openSUSE_Factory )

**Comment by copyme** [comment]
@luziferius Unfortunately, the patch does not work. I just had a situation with two autokey-qt processes after re-login. One of them was allocating memory like a crazy, etc.

**Comment by luziferius** [comment]
I’ve pushed a somewhat hacky fix. AutoKey will exit forcefully, when it receives a `Xlib.error.ConnectionClosedError`.
This method may have unintended consequences, but gets the job done.

Edit: Released version [`0.95.5`](https://github.com/autokey/autokey/releases/tag/v0.95.5), containing the bug fix.
